logstash/lib/logstash/outputs/loki.rb

```diff
@@ -0,0 +1,268 @@
+  config :url, :validate => :string,  :required => true
+
+  ## 'BasicAuth credentials'
+  config :username, :validate => :string, :required => false
+  config :password, :validate => :string, secret: true, :required => false
+
+  ## 'Client certificate'
+  config :cert, :validate => :path, :required => false
+  config :key, :validate => :path, :required => false
+
+  ## 'TLS'
+  config :ca_cert, :validate => :path, :required => false
+
+  ## 'Loki Tenant ID'
+  config :tenant_id, :validate => :string, :default => "fake", :required => false
+
+  ## 'Maximum batch size to accrue before pushing to Loki. Defaults to 102400 bytes'
+  config :batch_size, :validate => :number, :default => 102400, :required => false
+
+  ## 'Interval in seconds to wait before pushing a batch of records to Loki. Defaults to 1 second'
+  config :batch_wait, :validate => :number, :default => 1, :required => false
+
+  ## 'Array of label names to exclude from sending to Loki'
+  config :exclude_labels, :validate => :array, :default => [], :required => false 
+
+  ## 'Extra labels to add to all log streams'
+  config :external_labels, :valide => :hash,  :default => {}, :required => false
+
+  ## 'Log line to field to pick from logstash. Defaults to "message"'
+  config :message_field, :validate => :string, :default => "message", :required => false
+
+  public
+  def register
+    @uri = URI.parse(@url + '/loki/api/v1/push')
+    unless @uri.is_a?(URI::HTTP) || @uri.is_a?(URI::HTTPS)
+      raise LogStash::ConfigurationError, "url parameter must be valid HTTP, currently '#{@url}'"
+    end
+
+    @logger.info("New Loki output", :class => self.class.name)
+
+    @Channel = Concurrent::Channel
+    @entries = @Channel.new
+    @exclude_labels += ["message", "@timestamp"] 
+    @retry_http = 3
+    @delay = 1
+    @batch = nil
+
+    if ssl_cert?
+      load_ssl
+      validate_ssl_key
+    end
+
+    @Channel.go{run()}
+  end
+
+  def ssl_cert?
+    !@key.nil? && !@cert.nil?
+  end
+
+  def load_ssl
+    @cert = OpenSSL::X509::Certificate.new(File.read(@cert)) if @cert
+    @key = OpenSSL::PKey.read(File.read(@key)) if @key
+  end
+
+  def validate_ssl_key
+    if !@key.is_a?(OpenSSL::PKey::RSA) && !@key.is_a?(OpenSSL::PKey::DSA)
+      raise "Unsupported private key type #{key.class}"
+    end
+  end
+
+  def ssl_opts(uri)
+    opts = {
+      use_ssl: uri.scheme == 'https'
+    }
+
+    if !@cert.nil? && !@key.nil?
+      opts = opts.merge(
+        verify_mode: OpenSSL::SSL::VERIFY_PEER,
+        cert: @cert,
+        key: @key
+      )
+    end
+
+    unless @ca_cert.nil?
+      opts = opts.merge(
+        ca_file: @ca_cert
+      )
+    end
+    opts
+  end 
+
+  def run()
+	  min_wait_checkfrequency = 1/1000 #1 millisecond
+	  max_wait_checkfrequency = @batch_wait
+	  if max_wait_checkfrequency < min_wait_checkfrequency
+		  max_wait_checkfrequency = min_wait_checkfrequency
+    end
+
+    max_wait_check = Concurrent::Channel.tick(max_wait_checkfrequency)
```

Comment:
I think this is Ticker and not tick see https://github.com/ruby-concurrency/concurrent-ruby/blob/master/examples/go-by-example-channels/ticker.rb

I think you also need to cleanly stop it. Although I don't think it has side effects.

Reply:
tick is an alias for ticker. We can use either of the two.
**FYR:** `Default Selection` at last https://github.com/ruby-concurrency/concurrent-ruby/blob/master/docs-source/channel.md

Reply:
Yep nice, thanks for the info. I still think you need to gracefully stop it ?

```
ticker = Concurrent::Channel.ticker(0.5)
Concurrent::Channel.go do
  ticker.each do |tick|
    print "Tick at #{tick}\n"
  end
end

sleep(1.6)
ticker.stop # alias for `close`
print "Ticker stopped\n"
```
`ticker.stop` seems to do the job.